xc/domain.ml

```diff
@@ -1903,3 +1903,51 @@ let set_xsdata ~xs domid xsdata =
       List.iter (fun x -> t.Xst.rm (dom_path ^ "/" ^ x)) allowed_xsdata_prefixes;
       t.Xst.writev dom_path (filtered_xsdata xsdata);
     )
+
+type node =
+  { contents: string;
+    subtrees: (string * node) list }
+
+let move_xstree ~xs domid olduuid newuuid =
+  let search_paths = [[""; "local"; "domain"; string_of_int domid];
+                      [""; "xapi"; olduuid];
+                      [""; "vss"; olduuid];
+                      [""; "vm"; olduuid]] in
+
+  let regexp = Re_pcre.regexp olduuid in
+
+  let rec get_tree t path =
+    let open Xenstore in
+    let subtrees = try t.Xs.directory (String.concat "/" path) with Xs_protocol.Invalid -> [] in
+    let subtrees = subtrees |> List.filter (fun s -> s  <> "") in
+    let contents = t.Xs.read (String.concat "/" path) in
+    { contents;
+      subtrees = List.map (fun f -> (f, get_tree t (path @ [f]))) subtrees } in
+
+  let mv_tree path =
+    let open Xenstore in
+    Xs.transaction xs (fun t ->
```

Comment:
Right now I am glad that we have transactions in xenstore. Is the impact of such a large rewrite tangible?

Reply:
If you mean from a performance point of view there is a small noticable upwards trend (slower) on the vm migration graphs, but it is well within the noise of the measurement itself. I think for SMAPIv1 the SR level operations dominate the migration downtime.